txnsync/mainloop.go

```diff
@@ -0,0 +1,436 @@
+		// the transaction pool is currently full.
+		if float32(ent.transactionPoolSize) < float32(s.config.TxPoolSize)*transactionPoolLowWatermark {
+			s.transactionPoolFull = false
+		}
+	} else {
+		if float32(ent.transactionPoolSize) > float32(s.config.TxPoolSize)*transactionPoolHighWatermark {
+			s.transactionPoolFull = true
+		}
+	}
+
+	newBeta := beta(ent.transactionPoolSize)
+
+	// check if beta should be updated
+	if !shouldUpdateBeta(s.lastBeta, newBeta, betaGranularChangeThreshold) {
+		// no changes
+		return
+	}
+	// yes, change beta as the number of transactions in the pool have changed dramatically since the last time.
+	s.lastBeta = newBeta
+
+	peers := make([]*Peer, 0, len(s.interruptablePeers))
+	for _, peer := range s.interruptablePeers {
+		if peer == nil {
+			continue
+		}
+		peers = append(peers, peer)
+		peer.state = peerStateHoldsoff
+	}
+
+	// reset the interruptablePeers array, since all it's members were made into holdsoff
+	s.interruptablePeers = nil
+	s.interruptablePeersMap = make(map[*Peer]int)
+	deadlineMonitor := s.clock.DeadlineMonitorAt(s.clock.Since() + sendMessagesTime)
+	s.sendMessageLoop(s.clock.Since(), deadlineMonitor, peers)
+
+	currentTimeout := s.clock.Since()
+	for _, peer := range peers {
+		peerNext := s.scheduler.peerDuration(peer)
+		if peerNext < currentTimeout {
+			// shouldn't be, but let's reschedule it if this is the case.
+			s.scheduler.schedulerPeer(peer, currentTimeout+s.lastBeta)
+			continue
+		}
+		// given that peerNext is after currentTimeout, find out what's the difference, and divide by the beta.
+		betaCount := (peerNext - currentTimeout) / s.lastBeta
+		peerNext = currentTimeout + s.lastBeta*betaCount
+		s.scheduler.schedulerPeer(peer, peerNext)
+	}
+}
+
+// calculate the beta parameter, based on the transaction pool size.
+func beta(txPoolSize int) time.Duration {
+	if txPoolSize < 200 {
+		txPoolSize = 200
+	} else if txPoolSize > 10000 {
+		txPoolSize = 10000
+	}
+	beta := 1.0 / (2 * 3.6923 * math.Exp(float64(txPoolSize)*0.00026))
+	return time.Duration(float64(time.Second) * beta)
+
+}
+
+func shouldUpdateBeta(currentBeta, newBeta time.Duration, betaGranularChangeThreshold float32) bool {
+	// see if the newBeta is at least threshold percent smaller or bigger than the current one
+	if float32(newBeta) >= (float32(currentBeta) * (1.0 + betaGranularChangeThreshold)) {
+		return true
+	}
+	if float32(newBeta) <= (float32(currentBeta) * (1.0 - betaGranularChangeThreshold)) {
+		return true
+	}
+	// no, it's not.
+	return false
+}
+
+func (s *syncState) onNewRoundEvent(ent Event) {
+	s.clock = s.clock.Zero().(timers.WallClock)
+	peers := s.getPeers()
+	newRoundPeers := peers
+	if s.isRelay {
+		// on relays, outgoing peers have a difference scheduling, which is based on the incoming message timing
+		// rather then a periodic message transmission.
+		newRoundPeers = incomingPeersOnly(newRoundPeers)
+	}
+	s.scheduler.scheduleNewRound(newRoundPeers, s.isRelay)
```

Comment:
yes, that's a good point. what if.. we were to flip it so that we'll drop the test from `scheduleNewRound` ? 
The reason I'm suggesting this is that the filtering would be skipped on non-relays, and would be carried out on relays at a single location.

Reply:
I agree with that too. My issue is having a redundant check, and I understand you are proposing eliminating the other check. I think it is a great.

Reply:
updated.